File: deploy/post_process.cpp
#include <iostream>
#include <opencv2/opencv.hpp>
#include <vector>
#include <algorithm>

// Function to post-process the score map
std::vector<cv::Rect> post_process(const cv::Mat& score_map, double bin_thresh = 0.5, double pos_thresh = 0.3) {
    // Binary thresholding
    cv::Mat binary_mask = (score_map > bin_thresh) / 255;

    // Label connected components
    cv::Mat labeled_image, stats, centroids;
    int num_objects = cv::connectedComponentsWithStats(binary_mask, labeled_image, stats, centroids);

    // Calculate confidence scores for each bounding box
    std::vector<cv::Rect> bounding_boxes;
    for (int i = 1; i < num_objects; ++i) {
        int left = stats.at<int>(i, cv::CC_STAT_LEFT);
        int top = stats.at<int>(i, cv::CC_STAT_TOP);
        int width = stats.at<int>(i, cv::CC_STAT_WIDTH);
        int height = stats.at<int>(i, cv::CC_STAT_HEIGHT);

        cv::Mat mask = (labeled_image == i);
        cv::Scalar mean_score = cv::mean(score_map, mask);

        // Calculate the confidence score for the bounding box
        double confidence = mean_score[0];

        // Check if the confidence score is above the positive threshold
        if (confidence > pos_thresh) {
            bounding_boxes.push_back(cv::Rect(left, top, width, height));
        }
    }

    return bounding_boxes;
}

int main() {
    // Load your score map here (score_map should be a grayscale image)
    cv::Mat score_map = cv::imread("score_map.jpg", cv::IMREAD_GRAYSCALE);

    // Define binary and positive thresholds
    double bin_thresh = 0.5;
    double pos_thresh = 0.3;

    // Perform post-processing
    std::vector<cv::Rect> positive_boxes = post_process(score_map, bin_thresh, pos_thresh);

    // Output the positive bounding boxes
    for (const cv::Rect& box : positive_boxes) {
        std::cout << "Positive Bounding Box: (" << box.x << ", " << box.y << ", " << box.width << ", " << box.height << ")\n";
    }

    return 0;
}

File: deploy/post_process.py
import numpy as np
from skimage import measure

class Detection():
    def __init__(self, prob,  x1, y1, x2, y2, index=None, label=None,):
        self.label = label
        self.prob = prob
        self.x1 = x1
        self.x2 = x2
        self.y1 = y1
        self.y2 = y2

def post_process(score_map, bin_thresh=0.5, pos_thresh=0.3):
    # 二值化
    binary_mask = score_map > bin_thresh

    # 使用形态学操作将二值掩码转换为实体边界框
    labeled_image, num_objects = measure.label(binary_mask, connectivity=2, return_num=True)
    regions = measure.regionprops(labeled_image)
    bounding_boxes = [list(region.bbox) for region in regions]

    # 计算每个bounding box的置信度，等于scoremap*mask之后，在框内区域的有效平均值
    confidences = []
    for i, region in enumerate(regions):
        mask = labeled_image == i + 1
        masked_scores = score_map * mask
        masked_scores_sum = np.sum(masked_scores)
        masked_area = np.sum(mask)
        if masked_area > 0:
            confidence = masked_scores_sum / masked_area
            confidences.append(confidence)
        else:
            confidences.append(0.0)

    # # 根据置信度和pos_thresh选出阳性框
    positive_confidences = [confidence for confidence in confidences if confidence > pos_thresh]
    positive_boxes = [box for box, confidence in zip(bounding_boxes, confidences) if confidence > pos_thresh]
    return positive_confidences, positive_boxes

# 用法示例
score_map = np.random.rand(100, 100)  # 示例随机分数图
positive_boxes = post_process(score_map, bin_thresh=0.5, pos_thresh=0.3)
print("Positive Bounding Boxes:", positive_boxes)